components/timezone/src/provider/names.rs

```diff
@@ -0,0 +1,207 @@
+/// will sort into the same area, such that a map can be searched for both strict and loose equality.
+impl PartialOrd for NormalizedTimeZoneIdStr {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+/// The Ord impl will sort things using strict equality, but in such a way that all loose-equal items
+/// will sort into the same area, such that a map can be searched for both strict and loose equality.
+impl Ord for NormalizedTimeZoneIdStr {
+    fn cmp(&self, other: &Self) -> Ordering {
+        let cmp = self.cmp_loose(other);
+        // When loose equality holds, fall back to strict equality
+        if cmp == Ordering::Equal {
+            self.0.cmp(&other.0)
+        } else {
+            cmp
+        }
+    }
+}
+
+impl NormalizedTimeZoneIdStr {
+    /// Perform the loose comparison as defined in [`NormalizedTimeZoneIdStr`].
+    pub fn cmp_loose(&self, other: &Self) -> Ordering {
+        let self_iter = self.0.iter().map(u8::to_ascii_lowercase);
+        let other_iter = other.0.iter().map(u8::to_ascii_lowercase);
+        self_iter.cmp(other_iter)
+    }
+
+    /// Convert a string reference to a [`NormalizedTimeZoneIdStr`].
+    pub const fn from_str(s: &str) -> &Self {
+        Self::cast_ref(UnvalidatedStr::from_str(s))
+    }
+
+    /// Convert a [`UnvalidatedStr`] reference to a [`NormalizedTimeZoneIdStr`] reference.
+    pub const fn cast_ref(value: &UnvalidatedStr) -> &Self {
+        // Safety: repr(transparent)
+        unsafe { core::mem::transmute(value) }
+    }
+
+    /// Convert a [`UnvalidatedStr`] box to a [`NormalizedTimeZoneIdStr`] box.
+    pub const fn cast_box(value: Box<UnvalidatedStr>) -> Box<Self> {
+        // Safety: repr(transparent)
+        unsafe { core::mem::transmute(value) }
+    }
+
+    /// Get a [`NormalizedPropertyName`] box from a byte slice.
+    pub fn boxed_from_bytes(b: &[u8]) -> Box<Self> {
+        Self::cast_box(UnvalidatedStr::from_boxed_bytes(b.into()))
+    }
+}
+
+/// A mapping from IANA time zone identifiers to BCP-47 time zone identifiers.
+///
+/// Multiple IANA time zone IDs can map to the same BCP-47 time zone ID.
+///
+/// <div class="stab unstable">
+/// 🚧 This code is considered unstable; it may change at any time, in breaking or non-breaking ways,
+/// including in SemVer minor releases. While the serde representation of data structs is guaranteed
+/// to be stable, their Rust representation might not be. Use with caution.
+/// </div>
+#[derive(Debug, Clone)]
+#[icu_provider::data_struct(marker(IanaToBcp47MapV1Marker, "time_zone/iana_to_bcp47@1"))]
+#[cfg_attr(
+    feature = "datagen", 
+    derive(serde::Serialize, databake::Bake),
+    databake(path = icu_timezone::provider::names),
+)]
+#[cfg_attr(feature = "serde", derive(serde::Deserialize))]
+#[yoke(prove_covariance_manually)]
+pub struct IanaToBcp47MapV1<'data> {
```

Comment:
Having both IanaToBcp and BcpToIana seems wasteful, it's the same data in different order, and it's not small data. How about two ZeroMaps where the value is the index of the key in the other zeromap?

Reply:
1. It's not exactly the same data; BCP47-to-IANA only contains canonical IDs, whereas IANA-to-BCP47 includes legacy and alias IANA IDs. Note that one is about 23% smaller than the other.
2. By far the primary use case is IANA-to-BCP47. Most clients don't need BCP47-to-IANA, but it may be needed to support things like the Temporal API. Everyone else can slice out the other 12 kB.
3. This follows the same design pattern as property names, which have separate keys for the two directions of mappings
4. Making it a single data struct would increase data size for the common case (but I'm not sure by how much). Your proposed data model works well with ZeroMap; but I was also thinking that I would like to move these to AsciiTrie which doesn't support random indexing.

Would you like to discuss further?

Reply:
Will AsciiTrie be ready by the time is is released, or will that be `V2` anyway?

Reply:
I'll experiment with some different data models, including experimental AsciiTrie, and report back the findings.

Reply:
Some data size findings:

- iana2bcp ZeroMap: 14475B (this PR)
- bcp2iana ZeroMap: 11249B (this PR)
- Combined iana/bcp with dual ZeroMaps: 15521B (Rob's suggestion)
- iana2bcp AsciiTrie: 9820B

Reply:
I think a 7% size increase is worth paying for getting the other key for free. AsciiTrie will need a new data marker anyway.

Reply:
Discussion: ok to go with AsciiTrie as long as it lands in time for 1.3